File: 23_11_week5/wangon/1058.cpp
#include <iostream>
#include <vector>
using namespace std;

int main()
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	int n;
	cin >> n;
	vector<vector<int> > direct(n, vector<int>(n, 0));
	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < n; j++)
		{
			char c;
			cin >> c;
			if (c == 'Y')
				direct[i][j] = 1;
		}
	}
	vector<vector<int> > indirect = direct;
	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < n; j++)
		{
			for (int k = 0; k < n; k++)
			{
				if (i != j && direct[i][j] == 0)
				{
					if (direct[j][k] == 1 && direct[k][i] == 1)
						indirect[i][j] = 1;
				}
			}
		}
	}
	int ans = 0;
	for (int i = 0; i < n; i++)
	{
		int friends = 0;
		for (int j = 0; j < n; j++)
		{
			if (indirect[i][j] == 1)
				friends++;
		}
		ans = max(ans, friends);
	}
	cout << ans << endl;
	return 0;
}

File: src/23_11_week5/wangon/1058.py
n = int(input())
direct = [[0] * n for _ in range(n)]

for i in range(n):
    row = input().strip()
    for j in range(n):
        if row[j] == 'Y':
            direct[i][j] = 1

indirect = [row[:] for row in direct]
for i in range(n):
    for j in range(n):
        for k in range(n):
            if i != j and direct[i][j] == 0:
                if direct[j][k] == 1 and direct[k][i] == 1:
                    indirect[i][j] = 1

ans = 0
for i in range(n):
    count = 0
    for j in range(n):
        if indirect[i][j] == 1:
             count += 1
    friends = count
    ans = max(ans, friends)
print(ans)

